code/modules/wiremod/components/utility/typecast.dm

```diff
@@ -27,33 +32,86 @@
 		PORT_TYPE_STRING,
 		PORT_TYPE_NUMBER,
 		PORT_COMPOSITE_TYPE_LIST,
+		PORT_COMPOSITE_TYPE_ASSOC_LIST,
 		PORT_TYPE_ATOM,
+		PORT_TYPE_TABLE,
 	)
+
 	typecast_options = add_option_port("Typecast Options", component_options)
 
-/obj/item/circuit_component/typecast/pre_input_received(datum/port/input/port)
-	var/current_option = typecast_options.value
-	if(current_type != current_option)
-		current_type = current_option
-		output_value.set_datatype(current_type)
+	list_type_options = add_option_port("List Type", (GLOB.wiremod_basic_types) - PORT_TYPE_SIGNAL)
 
+/obj/item/circuit_component/typecast/pre_input_received(datum/port/input/port)
+	update_output()
 
 /obj/item/circuit_component/typecast/input_received(datum/port/input/port)
 	var/current_option = typecast_options.value
 	var/value = input_value.value
 	var/value_to_set = null
-	switch(current_option)
-		if(PORT_TYPE_STRING)
-			if(istext(value))
-				value_to_set = value
-		if(PORT_TYPE_NUMBER)
-			if(isnum(value))
-				value_to_set = value
-		if(PORT_COMPOSITE_TYPE_LIST)
-			if(islist(value))
-				value_to_set = value
-		if(PORT_TYPE_ATOM)
-			if(isatom(value))
+
+	if (current_option == PORT_COMPOSITE_TYPE_LIST || current_option == PORT_COMPOSITE_TYPE_ASSOC_LIST)
+		if(islist(value))
+			var/is_correct_format = TRUE
+			var/is_correctly_typed = TRUE
+			var/want_assoc = (current_option == PORT_COMPOSITE_TYPE_ASSOC_LIST)
+
+			for (var/key as anything in value)
+				if (!isnull(value[key]))
+					is_correct_format = want_assoc
+					break
+
+			if (is_correct_format)
+				for (var/entry as anything in value)
+					if (want_assoc)
+						entry = value[entry]
+					if (!entry_valid(entry, list_type_options.value))
+						is_correctly_typed = FALSE
+						break
+
+			if (is_correct_format && is_correctly_typed)
```

Comment:
honestly, I just think you should keep this to primitive types only. You shouldn't ever need to typecast a composite type and I really don't want to encourage people muddling up their variables. In a good circuit, you shouldn't ever need to typecast.